metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java

```diff
@@ -0,0 +1,491 @@
+     * @param metadataOffset    The latest metadata offset of the broker.
+     */
+    void touch(int brokerId, boolean fenced, long metadataOffset) {
+        BrokerHeartbeatState broker = brokers.get(brokerId);
+        if (broker == null) {
+            broker = new BrokerHeartbeatState(brokerId);
+            brokers.put(brokerId, broker);
+        } else {
+            if (!broker.fenced()) {
+                unfenced.remove(broker);
+                if (!broker.shuttingDown()) {
+                    active.remove(broker);
+                }
+            }
+        }
+        broker.lastContactNs = time.nanoseconds();
+        broker.metadataOffset = metadataOffset;
+        if (fenced) {
+            broker.shutdownOffset = -1;
+        } else {
+            unfenced.add(broker);
+            if (!broker.shuttingDown()) {
+                active.add(broker);
+            }
+        }
+    }
+
+    void beginBrokerShutDown(int brokerId, boolean deferred) {
+        BrokerHeartbeatState broker = brokers.get(brokerId);
+        if (broker == null) {
+            throw new RuntimeException("Unable to locate broker " + brokerId);
+        }
+        if (!broker.shuttingDown()) {
+            log.debug("broker {} has entered controlled shutdown.", brokerId);
+            active.remove(broker);
+            broker.shutdownOffset = deferred ? Long.MAX_VALUE : 0;
+        }
+    }
+
+    boolean shouldShutDown(int brokerId) {
+        BrokerHeartbeatState broker = brokers.get(brokerId);
+        if (broker == null) {
+            throw new RuntimeException("Unable to locate broker " + brokerId);
+        }
+        if (broker.fenced()) {
+            log.debug("broker {} can shut down immediately because it is fenced.");
+            return true;
+        }
+        if (!broker.shuttingDown()) {
+            log.trace("broker {} should not shut down because it is not in the " +
+                "shutdown state.", brokerId);
+            return false;
+        }
+        long curOffset = lowestActiveOffset();
+        if (curOffset >= broker.shutdownOffset) {
+            log.debug("broker {} should shut down. The current lowest active offset, {} " +
+                "is greater or equal to the broker shutdown offset of {}", brokerId,
+                curOffset, broker.shutdownOffset);
+            return true;
+        } else {
+            log.debug("broker {} must wait to shut down. The current lowest active offset, {} " +
+                    "is lower than the broker shutdown offset of {}", brokerId,
+                curOffset, broker.shutdownOffset);
+            return false;
+        }
+    }
+
+    long lowestActiveOffset() {
+        Iterator<BrokerHeartbeatState> iterator = active.iterator();
+        if (!iterator.hasNext()) {
+            return Long.MAX_VALUE;
+        }
+        BrokerHeartbeatState first = iterator.next();
+        return first.metadataOffset;
+    }
+
+    void updateShutdownOffset(int brokerId, long offset) {
```

Comment:
What's the difference between `beginBrokerShutDown()` and `updateShutdownOffset()`?  Why would the offset at which it can shutdown change?  A comment would be helpful.